src/utils/narrow.js

```diff
@@ -248,19 +248,24 @@ export const canSendToNarrow = (narrow: Narrow): boolean =>
     search: () => false,
   });
 
-/** True just if `haystack` contains all possible messages in `needle`. */
-export const narrowContains = (haystack: Narrow, needle: Narrow): boolean => {
-  if (isHomeNarrow(haystack)) {
-    return true;
-  }
-  if (isAllPrivateNarrow(haystack) && isPrivateOrGroupNarrow(needle)) {
-    return true;
-  }
-  if (isStreamNarrow(haystack) && needle[0].operand === haystack[0].operand) {
-    return true;
-  }
-  return JSON.stringify(needle) === JSON.stringify(haystack);
-};
+export const narrowContainsOutbox = (haystack: Narrow, needle: Outbox): boolean =>
+  caseNarrowPartial(haystack, {
+    stream: name => needle.type === 'stream' && needle.display_recipient === name,
+    topic: (streamName, topic) =>
+      needle.type === 'stream'
+      && needle.display_recipient === streamName
+      && needle.subject === topic,
+    pm: emails => emails === needle.display_recipient.map(r => r.email).join(','),
```

Comment:
This line (`pm: emails => ...`) doesn't look like it will work, for multiple reasons:
* It'll fail if `narrow[0].operand` and `display_recipient` are not in the same order. (In practice, we seem to frequently rely on serializations being consistent. :disappointed:)
* If one of your comments in a later commit is accurate – and, more questionably, if I'm understanding the code correctly –  the left-hand side will lack the self-user, while the right-hand side will have it.
* Both of the above bug(?)s are concealed by the fact that there's no `groupPm` case in this `caseNarrowPartial`. (I'd suggest using `caseNarrow` here instead.)
* It's a `TypeError` when `needle.type === 'stream'`, as `"stream name".map is not a function`.

Reply:
Eek, good catch re: missing the `groupPm` case!

I suspect what happened there is that this change (and the others in this branch) was originally taken from a longer `Narrow`-refactoring series where it came after a refactor to unify the `pm` and `groupPm` cases. I have a couple of longer `Narrow`-refactoring series as drafts, and I'm pretty sure I made that unification somewhere in there. Definitely `caseNarrow`, the total version, is the right one to use here for just that sort of reason; I have no recollection of what I might have been thinking when I wrote `caseNarrowPartial` instead.


> * It'll fail if `narrow[0].operand` and `display_recipient` are not in the same order. (In practice, we seem to frequently rely on serializations being consistent. disappointed)

Yeah, and I believe that's already the case in the existing version of this code, as a result of just string-comparing the results of `JSON.stringify`. Given that (which I'll re-check empirically, because although I think I did so originally, it's been some months since then and I'm not certain I did), I'm OK with preserving that behavior in the course of this basically orthogonal refactor. Probably deserves a comment about it, though.


> * If … the left-hand side will lack the self-user, while the right-hand side will have it.
> * It's a `TypeError` when `needle.type === 'stream'`, as `"stream name".map is not a function`.

Hmm. Will check on these. And probably write some unit tests for this function, as I seem to have demonstrated the need for them >_>

Reply:
> `JSON.stringify` … which I'll re-check empirically

Ah, no, here it is more directly:
* In the old code, we're comparing the `JSON.stringify` of each narrow, which means the sequences of operators and operands must be equal. In particular, `narrow[0].operand` must be equal.
* The `display_recipient` on an outbox message is constructed from the same `Narrow` value (passed to `addToOutbox`) that in the old code was included as a property. A quick look at the private helpers of `addToOutbox` shows that it's computed like so:
```js
  narrow[0].operand
    .split(',')
    .map(item => {
      const user = usersByEmail.get(item) || NULL_USER;
      return { email: item, id: user.user_id, full_name: user.full_name };
    })
    .concat({ email: selfDetail.email, id: selfDetail.user_id, full_name: selfDetail.full_name });
```
So yep, we've been counting on the order of these.

I think that same look at the code also confirms that the display_recipient will have the self user added at the end, like you said, while the narrow's list of emails won't.

Reply:
In #4299 I ended up dropping this implementation entirely... because there's already an `isMessageInNarrow`! Which I noticed partly as a consequence of https://github.com/zulip/zulip-mobile/pull/3889#discussion_r377400072 above. That function needed some significant work to become well tested, and then that testing turned up some bugs which needed fixing; those changes were #4294, and #4299 returns to the outbox code to switch to using that.